### PR TITLE
[SPARK-28093][SQL] Fix TRIM/LTRIM/RTRIM function parameter order issue

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

This pr fix `TRIM`/`LTRIM`/`RTRIM` function parameter order issue, otherwise:

```sql
spark-sql> SELECT trim('yxTomxx', 'xyz'), trim('xxxbarxxx', 'x');
z
spark-sql> SELECT ltrim('zzzytest', 'xyz'), ltrim('xyxXxyLAST WORD', 'xy');
xyz
spark-sql> SELECT rtrim('testxxzx', 'xyz'), rtrim('TURNERyxXxy', 'xy');
xy
spark-sql>
```

```sql
postgres=# SELECT trim('yxTomxx', 'xyz'), trim('xxxbarxxx', 'x');
 btrim | btrim
-------+-------
 Tom   | bar
(1 row)

postgres=# SELECT ltrim('zzzytest', 'xyz'), ltrim('xyxXxyLAST WORD', 'xy');
 ltrim |    ltrim
-------+--------------
 test  | XxyLAST WORD
(1 row)

postgres=# SELECT rtrim('testxxzx', 'xyz'), rtrim('TURNERyxXxy', 'xy');
 rtrim |   rtrim
-------+-----------
 test  | TURNERyxX
(1 row)
```

## How was this patch tested?

unit tests

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/inputs/string-functions.sql [hunk 1]
```diff
@@ -38,3 +38,11 @@ SELECT substring('Spark SQL', 5, 1);
 SELECT substring('Spark SQL' from 5);
 SELECT substring('Spark SQL' from -3);
 SELECT substring('Spark SQL' from 5 for 1);
+
+-- trim/ltrim/rtrim
```
Comment (wangyum) [inline sql/core/src/test/resources/sql-tests/inputs/string-functions.sql]: `StringExpressionsSuite` can not test these code paths. So add the test to `string-functions.sql`:
https://github.com/apache/spark/pull/24902/files#diff-2eb730f0723800778689468764627032R656
https://github.com/apache/spark/pull/24902/files#diff-2eb730f0723800778689468764627032R756
https://github.com/apache/spark/pull/24902/files#diff-2eb730f0723800778689468764627032R859

**Comment by wangyum** [comment]
retest this please

File: sql/core/src/test/resources/sql-tests/inputs/string-functions.sql [hunk 2]
```diff
@@ -38,3 +38,11 @@ SELECT substring('Spark SQL', 5, 1);
...
+SELECT trim('yxTomxx', 'xyz');
+SELECT trim('xxxbarxxx', 'x');
+SELECT ltrim('zzzytest', 'xyz');
```
Comment (gengliangwang) [inline sql/core/src/test/resources/sql-tests/inputs/string-functions.sql]: Nit: add some characters of 'xyz' at the end

File: sql/core/src/test/resources/sql-tests/inputs/string-functions.sql [hunk 3]
```diff
@@ -38,3 +38,11 @@ SELECT substring('Spark SQL', 5, 1);
...
+SELECT ltrim('xyxXxyLAST WORD', 'xy');
+SELECT rtrim('testxxzx', 'xyz');
```
Comment (gengliangwang) [inline sql/core/src/test/resources/sql-tests/inputs/string-functions.sql]: Nit: add some characters of 'xyz' at the begining

**Review by gengliangwang — APPROVED** [review]
Nice catch!

File: sql/core/src/test/resources/sql-tests/inputs/string-functions.sql (hunk 2 above)
Comment (wangyum) [reply sql/core/src/test/resources/sql-tests/inputs/string-functions.sql]: Done

File: sql/core/src/test/resources/sql-tests/inputs/string-functions.sql (hunk 3 above)
Comment (wangyum) [reply sql/core/src/test/resources/sql-tests/inputs/string-functions.sql]: Done

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @wangyum and @gengliangwang .
Wow. Amazingly old bug.

Merged to master.

**Comment by dongjoon-hyun** [comment]
Could you make two backporting PRs for branch-2.4 and branch-2.3, @wangyum ?
There exist conflicts at `string-functions.sql` and its output.

**Comment by srowen** [comment]
I marked this as 'release notes' as it's possible some callers have reversed their arguments just to work around this.

**Comment by cloud-fan** [comment]
I don't see the specification of the TRIM function parameters order in the SQL standard, and I assume it's vendor-specific. I agree that most SQL systems use the new order, but it doesn't mean the old order is wrong.

More importantly, the SQL standard does define the TRIM syntax as `TRIM ( [ characters FROM ] string )`, and I think this is the thing we should follow. I'd suggest we revert this commit (how can users figure out the parameter order of a function is changed? TRIM is a common function and this can break many queries silently.). We can implement `TRIM ( [ characters FROM ] string )` in 3.1.

**Comment by Ngone51** [comment]
Opened #27540 to revert this.

**Comment by dongjoon-hyun** [comment]
@cloud-fan . SQL Syntax and function is not the same. Why do we need to follow that?
> More importantly, the SQL standard does define the TRIM syntax as TRIM ( [ characters FROM ] string ), and I think this is the thing we should follow.

**Comment by cloud-fan** [comment]
After checking the codebase, we already support `TRIM ( [ characters FROM ] string )`. My question is: how can we silently change the parameter of an existing SQL function? I don't think this is a correctness issue as the SQL standard doesn't say that the function signature have to be `trim(srcStr, trimStr)`.

I agree this was a mistake in the first place, but it has been released and I don't think we can just change it. cc @marmbrus

**Comment by dongjoon-hyun** [comment]
Yes. I'm not against for your decision. What I'm asking is that let's do this in a public manner if we make this as a policy.

In the recent PRs (like raising runtime exceptions by default), we already notice that the following is your criteria. Then, let's make that public.
> how can we silently change the parameter

If we cannot change something for backward compatibility even in **Major** version, that's a big decision in the community because it's stricter than `Semantic versioning`.

**Comment by marmbrus** [comment]
@dongjoon-hyun I agree that it would be helpful to come up with a more concrete API evolution rubric to help the Spark project continue to grow and help resolve these debates uniformly across a growing number of reviewers. That said, I don't think worrying about compatibility / upgradability is a big policy change for the Spark project.

As far back as the debate to release Spark 1.0, when we were first debating the use of Semantic versioning, Matei said:

> I know that some names are suboptimal, but I absolutely detest breaking APIs, config names, etc. I’ve seen it happen way too often in other projects (even things we depend on that are officially post-1.0, like Akka or Protobuf or Hadoop), and it’s very painful. I think that we as fairly cutting-edge users are okay with libraries occasionally changing, but many others will consider it a show-stopper. Given this, I think that any cosmetic change now, even though it might improve clarity slightly, is not worth the tradeoff in terms of creating an update barrier for existing users.

Given the above, I strongly believe that having empathy for users upgrading Spark has been part of the project's ethos since the beginning. Nobody is arguing the we *cannot* break APIs at a Major version. The argument is the benefits of changing the API need to outweigh the costs to existing users.

In my opinion, this change is particularly insidious, as both arguments are strings, you won't even get an analysis exception. As @cloud-fan points out, the answers to queries are just going to silently change. This could cause wrong results in production, possibly for a very long time before someone notices.

In contrast, while a different parameter ordering might be more intuitive given experience with other SQL engines, I would guess people writing new Spark jobs will quickly figure out the right way to use this API (by just trying both orders, or by looking up the docs).

**Comment by gengliangwang** [comment]
@marmbrus Thanks for looking into this.
I am now +1 with keeping the original parameter order and reverting this one.

**Comment by dongjoon-hyun** [comment]
What Matei said is `cosmetic change` instead of bug fixes, isn't it?
> Given this, I think that any cosmetic change now

@marmbrus and @gengliangwang . As I mentioned here (https://github.com/apache/spark/pull/24902#issuecomment-585042944), I have been respecting your decisions, but we need to discuss this officially in the mailing list to categorize this types. 

IIUC, the main idea is you don't want to allow any `silently change`, isn't it? If then, please make it official policy at this time. That's my request. It's helpful for the community to review the upcoming PRs.

**Comment by marmbrus** [comment]
Reordering function parameters to match another system, for a method that is otherwise working correctly, sounds exactly like a cosmetic change to me.

And as I pointed out, this has been discussed officially on the mailing list. I gave one example, but I can assure you this is not the only one.

Don't read just his specific example, but rather also understand the motivation he gives. The Spark project always has been concerned about unnecessary pain being inflicted on users during an upgrade. He encourages us to think about "the tradeoff in terms of creating an update barrier for existing users".

I'm also not saying we should *never* silently change behavior. However, in general, silent behavior changes are a big red flag to me. I think they are extra costly to users for the reasons listed above.

I believe @cloud-fan is working on collecting a bunch of examples so that he can propose a framework on the mailing list to make sure we evaluate these cases consistently.

**Comment by cloud-fan** [comment]
Thanks @marmbrus for pointing out that this is already discussed officially on the mailing list. I'm trying to build a clear document about how to deal with behavior changes and make it more smooth for users to upgrade. But this may take a while as I need to collect many examples.

@dongjoon-hyun shall we move forward to revert this kind of cosmetic changes from 3.0 first? I don't want to block 3.0 until I finish the document.

**Comment by dongjoon-hyun** [comment]
I also don't want to block 3.0 release. If you guys are reluctant to send it officially, let me summarize this discussion on the dev mailing address.

For the following, it seems that we have different ideas on that.
> Reordering function parameters to match another system, for a method that is otherwise working correctly, sounds exactly like a cosmetic change to me.

**Comment by marmbrus** [comment]
Sorry, can you explain how different parameter orderings for a function don't clearly fit the definition of a "cosmetic difference"? An implementation of the function can return the correct result independent of the ordering.

Is there something I'm missing?

**Pull request closed without merging**